fdbserver/BackupWorker.actor.cpp

```diff
@@ -67,46 +67,162 @@ struct VersionedMessage {
 	const Version startVersion;
 	const Optional<Version> endVersion; // old epoch's end version (inclusive), or empty for current epoch
-	const LogEpoch recruitedEpoch;
-	const LogEpoch backupEpoch;
+	const LogEpoch recruitedEpoch; // current epoch whose tLogs are receiving mutations
+	const LogEpoch backupEpoch; // the epoch workers should pull mutations
+	LogEpoch oldestBackupEpoch = 0; // oldest epoch that still has data on tLogs for backup to pull
 	Version minKnownCommittedVersion;
 	Version savedVersion;
 	AsyncVar<Reference<ILogSystem>> logSystem;
 	Database cx;
 	std::vector<VersionedMessage> messages;
-	AsyncVar<bool> pullFinished;
+	NotifiedVersion pulledVersion;
+	bool pulling = false;
+	bool stopped = false;
+	bool exitEarly = false; // If the worker is on an old epoch and all backups starts a version >= the endVersion
 
 	struct PerBackupInfo {
 		PerBackupInfo() = default;
-		PerBackupInfo(BackupData* data, Version v) : self(data), startVersion(v) {}
+		PerBackupInfo(BackupData* data, UID uid, Version v) : self(data), startVersion(v) {
+			// Open the container and get key ranges
+			BackupConfig config(uid);
+			container = config.backupContainer().get(data->cx);
+			ranges = config.backupRanges().get(data->cx);
+			if (self->backupEpoch == self->recruitedEpoch) {
+				// Only current epoch's worker update the number of backup workers.
+				updateWorker = _updateStartedWorkers(this, data, uid);
+			}
+			TraceEvent("BackupWorkerAddJob", data->myId).detail("BackupID", uid).detail("Version", v);
+		}
+
+		void stop() {
+			stopped = true;
+			updateWorker = Void(); // cancel actors
+		}
+
+		void cancelUpdater() { updateWorker = Void(); }
 
-		bool isRunning() {
-			return container.isReady() && ranges.isReady() && !stopped;
+		bool isReady() const {
+			return stopped || (container.isReady() && ranges.isReady());
+		}
+
+		Future<Void> waitReady() {
+			if (stopped) return Void();
+			return _waitReady(this);
+		}
+
+		ACTOR static Future<Void> _waitReady(PerBackupInfo* info) {
+			wait(success(info->container) && success(info->ranges));
+			return Void();
+		}
+
+		// Update the number of backup workers in the BackupConfig. Each worker
+		// writes (epoch, tag.id) into the key. Worker 0 monitors the key and once
+		// all workers have updated the key, this backup is considered as started
+		// (i.e., the "submitBackup" call is successful). Worker 0 then sets
+		// the "allWorkerStarted" flag, which in turn unblocks
+		// StartFullBackupTaskFunc::_execute.
+		ACTOR static Future<Void> _updateStartedWorkers(PerBackupInfo* info, BackupData* self, UID uid) {
+			state BackupConfig config(uid);
+			state Future<Void> watchFuture;
+			state bool updated = false;
+			state bool firstWorker = info->self->tag.id == 0;
+			state bool allUpdated = false;
+			state Optional<std::vector<std::pair<int64_t, int64_t>>> workers;
+
+			loop {
+				state Reference<ReadYourWritesTransaction> tr(new ReadYourWritesTransaction(self->cx));
```

Comment:
move outside the loop to avoid reset the onError backoff timer on each error

Reply:
Done.